Invalid warning on Mixins should come before declarations when having multiple includes

I am currently working with sass-lint 1.1.0

When using multiple mixins within a selector the rule for 'Mixins should come before declarations' starts throwing invalid warnings when using 3 or more. In another scenario the prepending selector can also cause this issue to throw a warning.

I have setup this little scss file to reproduce this issue:

```
@mixin foo($content) {
    content: $content;
}

@mixin bar {
    background-color: red;
}

@mixin hello {
    border-radius: 3px;
}


.beverages {

    .coffee {
        width: 100px;
        height: 100px;
    }    

    .tea {
        @include foo('yo');
        @include bar;
        @include hello;
        display: block;
        width: 100px;
        height: 100px;
    }
}
```

This snippet throws the following warning through the CLI
```
scss/mixin.scss
  24:9  warning  Mixins should come before declarations  mixins-before-declarations
```

This doesn't occur in the following cases:
1. When disabling the selector `.coffee` the warning is gone.
2. When disabling `@include hello;`. The other includes do not make a difference.

--- a/lib/rules/mixins-before-declarations.js
+++ b/lib/rules/mixins-before-declarations.js
@@ -11,35 +11,37 @@ module.exports = {
     ]
   },
   'detect': function (ast, parser) {
-    var result = [],
-        error;
+    var result = [];
+    var error;
 
-    ast.traverseByType('block', function (block) {
-      var lastDeclaration = null;
-      block.traverse(function (item, j) {
-        if (item.type === 'include') {
-          if (j > lastDeclaration && lastDeclaration !== null) {
-            item.forEach('simpleSelector', function (name) {
-              if (parser.options.exclude.indexOf(name.content[0].content) === -1) {
-                error = {
-                  'ruleId': parser.rule.name,
-                  'line': item.start.line,
-                  'column': item.start.column,
-                  'message': 'Mixins should come before declarations',
-                  'severity': parser.severity
-                };
-                result = helpers.addUnique(result, error);
-              }
-            });
-          }
+    ast.traverseByType('include', function (node, i, parent) {
+      var depth = 0;
+      var declarationCount = [depth];
+
+      parent.traverse( function (item) {
+        if (item.type === 'ruleset') {
+          depth++;
+          declarationCount[depth] = 0;
         }
         if (item.type === 'declaration') {
-          lastDeclaration = j;
+          declarationCount[depth]++;
+        }
+        else if (item.type === 'include') {
+          item.forEach('simpleSelector', function (name) {
+            if (parser.options.exclude.indexOf(name.content[0].content) === -1 && declarationCount[depth] > 0) {
+              error = {
+                'ruleId': parser.rule.name,
+                'line': item.start.line,
+                'column': item.start.column,
+                'message': 'Mixins should come before declarations',
+                'severity': parser.severity
+              };
+              result = helpers.addUnique(result, error);
+            }
+          });
         }
       });
-      lastDeclaration = null;
     });
-
     return result;
   }
 };

--- a/lib/rules/mixins-before-declarations.js
+++ b/lib/rules/mixins-before-declarations.js
@@ -23,7 +23,7 @@ module.exports = {
           depth++;
           declarationCount[depth] = 0;
         }
-        if (item.type === 'declaration') {
+        else if (item.type === 'declaration') {
           declarationCount[depth]++;
         }
         else if (item.type === 'include') {

--- a/lib/rules/mixins-before-declarations.js
+++ b/lib/rules/mixins-before-declarations.js
@@ -11,12 +11,12 @@ module.exports = {
     ]
   },
   'detect': function (ast, parser) {
-    var result = [];
-    var error;
+    var result = [],
+        error;
 
     ast.traverseByType('include', function (node, i, parent) {
-      var depth = 0;
-      var declarationCount = [depth];
+      var depth = 0,
+          declarationCount = [depth];
 
       parent.traverse( function (item) {
         if (item.type === 'ruleset') {

--- a/tests/main.js
+++ b/tests/main.js
@@ -133,11 +133,35 @@ describe('rule', function () {
   });
 
   //////////////////////////////
-  // Mixins Before DEclarations
+  // Mixins Before Declarations
   //////////////////////////////
   it('mixins before declarations', function (done) {
     lintFile('mixins-before-declarations.scss', function (data) {
-      assert.equal(4, data.warningCount);
+      assert.equal(5, data.warningCount);
+      done();
+    });
+  });
+
+  //////////////////////////////
+  // Mixins Before Declarations - overwrite
+  //////////////////////////////
+  it('mixins before declarations - excludes', function (done) {
+    lintFile('mixins-before-declarations.scss', {
+      'rules': {
+        'mixins-before-declarations': [
+          1,
+          {
+            'exclude': [
+              'test-again',
+              'waldo',
+              'mq',
+              'breakpoint'
+            ]
+          }
+        ]
+      }
+    }, function (data) {
+      assert.equal(0, data.warningCount);
       done();
     });
   });

--- a/tests/sass/mixins-before-declarations.scss
+++ b/tests/sass/mixins-before-declarations.scss
@@ -25,5 +25,25 @@
     content: 'where';
 
     @include waldo;
+
+  }
+
+  &__element {
+    @include element;
+    width: 100%;
+
+    @include mq(500px) {
+      content: 'mq';
+    }
+
+    @include waldo;
+
+    &--modifier {
+      @include foo('yo');
+      @include hello;
+      @include test;
+      height: 100px;
+      @include test-again;
+    }
   }
 }

